### PR TITLE
Fix signed digits issue (depends on #68)

### PR DESCRIPTION
**Pull request by dtebbs**
Addresses #69 
- [x] Enable checks for bls12-381 in `libff/algebra/scalar_multiplication/tests/test_multiexp.cpp`

### CODE REVIEW

**Review by vesselinux — APPROVED** [review]
Nice work! The fix is actually much more complex than I thought. I applied it to the bls12-381 branch and enabled the test_signed_digits and the test passes this time.

File: libff/algebra/fields/field_utils.tcc [hunk 1]
```diff
@@ -74,8 +85,83 @@ size_t field_get_digit(
...
+template<typename FieldT>
+size_t field_get_num_signed_digits(const size_t digit_size)
+{
+    // -1 in the field must have the largest number of 1's in the higher order
+    // elements of its binary representation.
+    const bigint<FieldT::num_limbs> minus_one = (-FieldT::one()).as_bigint();
+
+    // The naive approach is to compute:
+    //
+    //   ciel(FieldT::num_bits + 1 / digit_size)
```
Comment (AntoineRondelet) [inline libff/algebra/fields/field_utils.tcc]: ciel -> ceil?

File: libff/algebra/fields/field_utils.tcc [hunk 2]
```diff
@@ -74,8 +85,83 @@ size_t field_get_digit(
...
+        // This (and any previous values) are max_signed_value. Any lower-order
+        // digits which have the signed bit set will trigger an overflow up to
+        // the most-significant digit.
+    }
+
+    if (!final_overflow) {
+        return naive_num_digits;
+    } else {
+        return naive_num_digits + 1;
+    }
```
Comment (AntoineRondelet) [inline libff/algebra/fields/field_utils.tcc:162]: Why not remove the `else` part here and simply do:

```cpp
    if (!final_overflow) {
        return naive_num_digits;
    }
    
    return naive_num_digits + 1;
```
?

Or the other way around
```cpp
if (final_overflow) {
    return naive_num_digits + 1;
}
return naive_num_digits;
```

Comment (AntoineRondelet) [inline libff/algebra/fields/field_utils.tcc:162]: Why not remove the `else` part here and simply do:

```cpp
    if (!final_overflow) {
        return naive_num_digits;
    }
    
    return naive_num_digits + 1;
```
?

Or the other way around
```cpp
if (final_overflow) {
    return naive_num_digits + 1;
}
return naive_num_digits;
```

File: libff/algebra/fields/field_utils.tcc [hunk 3]
```diff
@@ -86,10 +172,10 @@ ssize_t field_get_signed_digit(
     //  carry_mask       1 0 0 0
     //  overflow_mask  1 0 0 0 0
     const size_t carry_mask = 1ull << (digit_size - 1);
-    const ssize_t overflow_mask = 1ll << digit_size;
+    const ssize_t overflow_mask = 1ull << digit_size;
     size_t carry = 0;
     ssize_t overflow = 0;
-    size_t digit;
+    ssize_t digit;
```
Comment (AntoineRondelet) [inline libff/algebra/fields/field_utils.tcc]: Hum I probably missed it, but why do you need to make `digit` signed here? Was that an issue with the `(digit - (carry * overflow_mask))` part of the returned value of this function?

Comment (AntoineRondelet) [inline libff/algebra/fields/field_utils.tcc:178]: Does `digit` need to be signed here?

**Comment by dtebbs** [comment]
> ciel -> ceil?

fixed.  thanks.

File: libff/algebra/fields/field_utils.tcc (hunk 3 above)
Comment (dtebbs) [reply libff/algebra/fields/field_utils.tcc]: No, there was no issue.  It seemed the correct type when I re-read this function, but in hindsight, I think the original `size_t` is most appropriate for everything up until the final signed value is computed.  Reverted.

File: libff/algebra/fields/field_utils.tcc (hunk 2 above)
Comment (dtebbs) [reply libff/algebra/fields/field_utils.tcc:162]: (repeated comment?)

Comment (dtebbs) [reply libff/algebra/fields/field_utils.tcc:162]: Removed the `else`.

Comment (AntoineRondelet) [reply libff/algebra/fields/field_utils.tcc:162]: Yes thanks - not sure what happened.

File: libff/algebra/fields/field_utils.tcc (hunk 1 above)
Comment (AntoineRondelet) [reply libff/algebra/fields/field_utils.tcc]: https://github.com/clearmatics/libff/pull/71#issuecomment-1094949770

File: libff/algebra/fields/field_utils.tcc (hunk 3 above)
Comment (AntoineRondelet) [reply libff/algebra/fields/field_utils.tcc]: Ah ok

**Comment by dtebbs** [comment]
(rebase onto `origin/bls12-381` at 87f4d925c98c95e49e0c9225e9c604ed7f434b4d)

**Comment by AntoineRondelet** [comment]
@dtebbs could you rebase on top of `develop` to clean the history now that `bls12-381` has been squashed-merged?

**Comment by dtebbs** [comment]
Thanks @AntoineRondelet.  I've just rebased and had a final check through the diffs.  All seems OK now.

**Review by AntoineRondelet — APPROVED** [review]
LGTM

**Pull request merged**